### DISCUSSION TITLE
Character get knocked up when runs into steep slopes

### DISCUSSION THREAD

**Discussion by demiurghg** [bepu/bepuphysics1#37] [Q&A] [answered]
Hi!

I have character controller (settings are similar to standard) and non-regular terrain-like (valley) static geometry. When character runs and hits steep slope character get knocked up pretty high.

I've suspected Vertical- and HorizontalMotionConstraints and CorrectContacts(), but experiments show that they does not affect such behavior.

![image](https://user-images.githubusercontent.com/3645657/224579198-30a5fd50-49d7-4e8f-b5c3-de6c110289e9.png)

Please, see video: https://youtu.be/b5spqg16MCk
See subtitles for description.

Could you suggest, how to eliminate or reduce such behavior?

**Comment by RossNordby** [comment, marked as answer]
The `CharacterController.VerticalMotionConstraint.MaximumGlueForce` tries to keep the character glued to the last support, which should stop this kind of thing from happening so long as the force is strong enough. You may need to increase it if the character's horizontal motion is overpowering it.

You can try out different values in the `CharacterPlaygroundDemo` and see their impact on the tilted planks (toward the upper left upon spawning). Reducing the maximum glue force lets the character slide up the wall like in the video.

**Reply by demiurghg** [reply]
Tried MaximumGlueForce = 50000. It improves situation in many cases, but when traction is lost (i.e. run forward --> jump --> hit the wall) character get knocked much higher than simple jump. 

Played a little and discover this effect does not occur when character hits steep slope running on horizontal surface. Instead, effects occurs when character move from gentle slope to steep slope or when airborne character hits the slope.

Checked the demo, everything works fine, character stop movement, when hit steep slope. Guess, such behavior occur when character already has small vertical velocity component before hitting the steep slope. 

Would you suggest area to investigate further?

**Reply by RossNordby** [reply]
>  when traction is lost (i.e. run forward --> jump --> hit the wall) character get knocked much higher than simple jump.

This is likely just a normal physical result.

A frictionless contact will attempt to subtract all penetrating velocity along the normal. Hitting a wall dead on subtracts away all the velocity, because the normal perfectly opposes it ('s' here refers to speed):
![image](https://user-images.githubusercontent.com/657209/224589128-fe939112-081d-432f-984f-feedac587b74.png)

Adding in an intermediate 45 degree slope means you lose cos(45 degrees) on each transition:
![image](https://user-images.githubusercontent.com/657209/224589210-ba82dfe3-1659-4582-8749-37a51ac3bb31.png)

As you smooth out the curve, more of the original speed can be preserved:
![image](https://user-images.githubusercontent.com/657209/224589288-fe63035f-37d6-4fc7-bfb7-fed70be5864a.png)

Most game characters tend to be *fast* relative to humans, so it's easy for horizontal velocity that gets redirected upwards to launch them.

> Instead, effects occurs when character move from gentle slope to steep slope or when airborne character hits the slope.

For the former (still supported) case: the more velocity the vertical motion constraint has to fight with its 'glue' effect, the higher its maximum force has to be to manage it. If the character's moving at very high speed on a slope that still has traction, and then transitions to a *slightly* steeper slope that doesn't have traction, it's extremely hard for the vertical motion constraint to stop all forward motion. The surface normal that the glue acts on would be almost entirely perpendicular to the forward motion at that point. Launching is probable.

Once a character loses traction (as in the airborne case), it's basically just behaving like a regular dynamic body, hence launching.

Or to phrase it another way: in the real world, your feet don't have glue, and if you ran at a gradually sloped wall, you'd be able to run up it too (and probably higher than you can jump). In a game where default physical behavior is not desired, you'll have to come up with exactly what it should do instead, and apply whatever nonphysical workarounds achieve it. The glue is one example of that.

If you don't want the redirection of jumping into a sloped wall when you don't have traction, you'll have to fight it somehow. One option: when the character lacks traction and finds contacts below it that don't provide traction, arbitrarily damp the upward momentum. (Take care with the implementation- you'll probably want to damp based on the *relative* velocity of the character and the thing it's hitting if you're, say, on a boat or a plane, because damping relative to the absolute frame of reference might slam the character into the floor or ceiling.)

**Reply by demiurghg** [reply]
Thank you for explanation.

I came to the following implementation. Code below is executed at the end of `void IBeforeSolverUpdateable.Update(float dt)`.
The idea is following: check side contacts, if they has no traction and not perfectly vertical, apply counter-impulse to reduce normal and positive vertical component of body's velocity.

```csharp
foreach (var c in SupportFinder.SideContacts)
{
    var v = Body.LinearVelocity;
    var n = c.Contact.Normal;

    // make normal and velocity planar:
    v.Y *= 0; 
    n.Y *= 0;

    var d = Vector3.Dot( n, v );

    var isTraction = Vector3.Dot( c.Contact.Normal, Body.orientationMatrix.Down ) < ContactCategorizer.TractionThreshold;
    var isVertical = Math.Abs( c.Contact.Normal.Y ) < 0.05f;

    if (d>0.001f && isTraction)
    {
        Vector3 impulse;

        // do not apply damping near perfectly vertical walls to allow jump:
        if (Body.LinearVelocity.Y > 0.01f && isVertical)
        {
            impulse = Math.Abs(Body.LinearVelocity.Y) * Vector3.Down * Body.Mass * 0.5f;
            Body.ApplyLinearImpulse( ref impulse );
        }

        var p = (-1) * Vector3.Dot(v,n) * n / n.Length();

        impulse = p * Body.Mass * 0.5f;
        Body.ApplyLinearImpulse( ref impulse );
    }
}
```
As result character does not rocket to the sky, but movement became little bit jerky especially near uneven walls.

Could you suggest any improvements or better place to execute this code? (relative velocity is kept in mind)

**Reply by RossNordby** [reply]
Not sure I understand the intent, but:
1. The impulse acting along Vector3.Down only activates when there's a side contact that's a roughly vertical surface, despite what the comment seems to imply.
2. Given the condition that the linear velocity is positive (>0.01), there's no need for the abs on the impulse calculation.
3. You can avoid computing an impulse and work directly on linear velocities. In fact, the `ApplyLinearImpulse` function is just:
```cs
linearVelocity.X += impulse.X * inverseMass;
linearVelocity.Y += impulse.Y * inverseMass;
linearVelocity.Z += impulse.Z * inverseMass;
```
4. The last impulse appears to be opposing part of the velocity along the contact normal projected onto the Y = 0 plane. I wouldn't expect this to do useful things unless you have some other goal.
5. Applying the damping only if there exists a support (or otherwise an active contact on the bottom end of the character) would avoid obstructing jumps against flat walls, while strongly fighting against launching up steep sloped surfaces.

**Reply by demiurghg** [reply]
4. The idea is to stop character from climbing by damping horizontal component of its velocity. This solution works in 90% cases but fails on very uneven walls.
5. As far I understand, the problem is caused by non-supportive contacts: i.e. run + jump + hit the wall = get knocked up, that why I prevent from vertical motion only when side contacts are not supportive.

The solution seems to work. Thank you for assistance!

**Reply by RossNordby** [reply]
> As far I understand, the problem is caused by non-supportive contacts: i.e. run + jump + hit the wall = get knocked up, that why I prevent from vertical motion only when side contacts are not supportive.

The problem is caused by contacts which do not provide *traction* but still have a normal capable of redirecting the character upward (by being a contact on the bottom with a normal pointing at least a little downward). By 5, I mean applying the damping if there is any *non-traction providing* contact still on the bottom of the character (a normal pointing at least somewhat downward).

Notably, the set of "supporting" contacts found by the character are a superset of traction contacts. Not all of them have traction.
